File: design-pattern-javascript/01-creacionales/03.2-abstract-factory.js
/**
 * ! Abstract Factory:
 * Es un patrón de diseño que permite crear familias de objetos relacionados
 * sin especificar sus clases concretas.
 *
 * En lugar de crear objetos individuales directamente,
 * creamos fábricas que producen un conjunto de objetos relacionados.
 *
 * * Es útil cuando necesitas crear objetos que son parte de una familia
 * * y quieres asegurarte de que estos objetos se complementen entre sí.
 *
 * https://refactoring.guru/es/design-patterns/abstract-factory
 */

/**
 * !Instrucciones:
 	1.Completen las Clases de Productos:
    •	ElectricCar debe implementar Vehicle y mostrar el mensaje "Ensamblando un auto eléctrico".
    •	GasCar debe implementar Vehicle y mostrar el mensaje "Ensamblando un auto de combustión".
    •	ElectricEngine debe implementar Engine y mostrar el mensaje "Arrancando motor eléctrico".
    •	GasEngine debe implementar Engine y mostrar el mensaje "Arrancando motor de combustión".

	2.	Completen las Clases de Fábricas:
    •	ElectricVehicleFactory debe crear un ElectricCar y un ElectricEngine.
    •	GasVehicleFactory debe crear un GasCar y un GasEngine.

	3. Prueben el Código:
	  •	Ejecuten el código para asegurarse de que cada fábrica produce el tipo correcto de vehículo y motor.

 */

class ElectricCar {
	assemble() {
		console.log("Ensamblando un auto eléctrico");
	}
}

class GasCar {
	assemble() {
		console.log("Ensamblando un auto de combustión");
	}
}

class ElectricEngine {
	start() {
		console.log("Arrancando motor eléctrico");
	}
}

class GasEngine {
	start() {
		console.log("Arrancando motor de combustible");
	}
}

class ElectricVehicleFactory {
	createVehicle() {
		return new ElectricCar();
	}
	createEngine() {
		return new ElectricEngine();
	}
}

class GasVehicleFactory {
	createVehicle() {
		return new GasCar();
	}
	createEngine() {
		return new GasEngine();
	}
}

function main(factory) {
	const vehicle = factory.createVehicle();
	const engine = factory.createEngine();

	vehicle.assemble();
	engine.start();
}

console.log("Creando vehículo eléctrico:");
main(new ElectricVehicleFactory());

console.log("\nCreando vehículo de combustión:");
main(new GasVehicleFactory());

File: design-pattern-python/01_creacionales/03.2_abstract_factory.py
from abc import ABC, abstractmethod

from rich.console import Console

console = Console()


class Vehicle(ABC):
    @abstractmethod
    def assemble(self) -> None:
        ...


class Engine(ABC):
    @abstractmethod
    def start(self) -> None:
        ...


class ElectricCar(Vehicle):
    def assemble(self) -> None:
        console.print('Ensamblando un auto [purple]eléctrico[/purple]')


class GasCar(Vehicle):
    def assemble(self) -> None:
        console.print('Ensamblando un auto de [gray30]combustion[/gray30]')


class ElectricEngine(Engine):
    def start(self) -> None:
        console.print('Arrancando motor [pink1]eléctrico[/pink1]')


class GasEngine(Engine):
    def start(self) -> None:
        console.print('Arrancando motor de [cyan]combustion[/cyan]')


class VehicleFactory(ABC):
    @abstractmethod
    def create_vehicle(self) -> Vehicle:
        ...

    @abstractmethod
    def create_engine(self) -> Engine:
        ...


class ElectricVehicleFactory(VehicleFactory):
    def create_vehicle(self) -> Vehicle:
        return ElectricCar()

    def create_engine(self) -> Engine:
        return ElectricEngine()


class GasVehicleFactory(VehicleFactory):
    def create_vehicle(self) -> Vehicle:
        return GasCar()

    def create_engine(self) -> Engine:
        return GasEngine()


def main(factory: VehicleFactory) -> None:
    vehicle = factory.create_vehicle()
    engine = factory.create_engine()

    vehicle.assemble()
    engine.start()


console.print('Creando vehículo eléctrico:')
main(ElectricVehicleFactory())

console.print('\nCreando vehículo de combustion:')
main(GasVehicleFactory())
